configure: build the debug and instrumented runtimes by default

gasche opened the pull request:
For a user that discovers a use for either of those, not having built
by default makes the experience pretty bad: they have to recompile
their own OCaml compiler, possibly from a source checkout if the
option is not available on opam...

With this change the total compilation time only increased by
7 seconds on my machine, from 3m24s to 3m33s -- these are sequential
builds, with parallel builds the difference is lost in the noise.

gasche committed:
build the debug and instrumented runtimes by default

For a user that discovers a use for either of those, not having built
by default makes the experience pretty bad: they have to recompile
their own OCaml compiler, possibly from a source checkout if the
option is not available on opam...

With this change the total compilation time only increased by
7 seconds on my machine, from 3m24s to 3m33s -- these are sequential
builds, with parallel builds the difference is lost in the noise.

shindere commented:
If you want to enable the debug runtime by default you may also want to
enable it in the Windows configurations.

For the instrumented runtime, though, things are not that simple.
Indeed, as far as I understood, this runtime relies on a system call
whihc is only available under Linux. I don't remember the details and
don't have time to figure them out right now but can investigate this
later. @damiendoligez would be able to explain more on this.

shindere commented:
@gasche in addition to my previous comment, look for "CAML_INSTR" in
byterun/misc.c, around line 200.

Once we have autoconf I think the right approach will be to automatically
detect whether the system is able to support this and adjust whether the
instrumented runtime is compiled accordingly.

gasche commented:
I propose to restrict the scope of this PR to changes in systems affected by the `./configure` script. I don't have a Windows machine to test with, so I would rather let someone else write a Windows-specific PR if they wish to.

dra27 commented:
It's a small tweak to config/Makefile.m* to build the debug runtimes on Windows, and Inria CI and AppVeyor can verify them... they were working last time I tested them

shindere commented:
Gabriel Scherer (2017/12/12 01:01 -0800):
> I propose to restrict the scope of this PR to changes in systems
> affected by the `./configure` script. I don't have a Windows machine
> to test with, so I would rather let someone else write a
> Windows-specific PR if they with to.

Keep in mind the instrumented runtime won't work on Mac OS or BSD, say.

shindere commented:
David Allsopp (2017/12/12 01:03 -0800):
> It's a small tweak to config/Makefile.m* to build the debug runtimes
> on Windows, and Inria CI and AppVeyor can verify them...

Inria's CI does so for the debugging runtime, yes.

gasche commented:
I don't want to work on Windows in this PR. I don't even know the file extension for `libcamlrun` there to mention in the Changes entry. Again, feel free to send a pull request.

I have to deal with the issue of whether setting the instrumented runtime by default on non-linux systems is correct. Maybe the simplest approach is to only enable it by default on Linux systems.

xavierleroy commented:
> I have to deal with the issue of whether setting the instrumented runtime by default on non-linux systems is correct. Maybe the simplest approach is to only enable it by default on Linux systems.

It's not a Linux / other-OS issue, it's a `clock_gettime` is available / is not available issue.  If you read the fine `configure` script, you'll see that configuring with instrumented runtime will fail if `clock_gettime` can't be found.  You could change the `configure` logic so that if `clock_gettime` is not found, no error is generated and `with_instrumented_runtime` is set to false.  But it would be nice to get confirmation from whomever wrote this test initially.

gasche commented:
(Whoever is probably @damiendoligez.)

Thanks, I'll have a look.

gasche commented:
In https://github.com/ocaml/ocaml/pull/1526/commits/ae8e4872c33df60d0fcde4a68306ab27b061390a, missing `clock_gettime` is an error only if `-with-instrumented-runtime` was passed explicitly by the user. In the default setting, if it is missing then the instrumented runtime is disabled without failure.

gasche committed:
configure: missing clock_gettime is a soft error for -with-instrumented-runtime

We wish to make -with-instrumented-runtime the default, but not break
the build on systems that do not support it -- when clock_gettime is
missing. This is done by tracking whether -with-instrumented-runtime
was set explicitly or is the default value; in the latter case,
a missing clock_gettime defaults to no instrumented runtime, without
raising an error at configure-time.

xavierleroy reviewed (approved):
LGTM.
